File: src/saaty.R
library(knitr)
library(kableExtra)

# Round while preserving sum
round_preserve_sum <- function(values, digits) {
  scale <- 10^digits
  scaled <- values * scale
  floored <- floor(scaled)
  missing_units <- round(scale - sum(floored))

  remainders <- (floored + 1 - scaled) / scaled
  indices <- order(remainders)

  result <- floored
  result[indices[1:missing_units]] <- result[indices[1:missing_units]] + 1
  return(result / scale)
}

# Calculate weights from Saaty matrix
calculate_weights <- function(matrix, method = "eigenvector", digits = 3) {
  actual_matrix <- ifelse(matrix < 0, 1 / (-matrix), matrix)
  n <- nrow(actual_matrix)

  if (method == "average") {
    norm_matrix <- sweep(actual_matrix, 2, colSums(actual_matrix), "/")
    weights <- rowMeans(norm_matrix)
    lambda_max <- sum(colSums(actual_matrix) * weights)
  } else if (method == "eigenvector") {
    eig <- eigen(actual_matrix)
    max_index <- which.max(Re(eig$values))
    weights <- Re(eig$vectors[, max_index])
    lambda_max <- Re(eig$values[max_index])
  } else {
    stop("Invalid method.")
  }

  weights <- weights / sum(weights)

  # Compute Consistency Index (CI)
  ci <- (lambda_max - n) / (n - 1)

  # Random Index (RI) values from Saaty for matrices of size 1–10
  ri_values <- c(0, 0, 0.58, 0.90, 1.12, 1.24, 1.32, 1.41, 1.45, 1.49)
  ri <- if (n <= length(ri_values)) ri_values[n] else NA

  cr <- if (!is.na(ri) && ri != 0) ci / ri else NA

  list(
    weights = round(weights, digits),
    consistency_ratio = round(cr, 3),
    lambda_max = round(lambda_max, 3),
    ci = round(ci, 3),
    ri = ri
  )
}


# Convert matrix to a data frame table
matrix_to_table <- function(matrix, attributes) {
  formatted <- apply(matrix, c(1, 2), function(x) if (x < 0) paste0("1/", -x) else as.character(x))
  df <- data.frame(Attribute = attributes, formatted, stringsAsFactors = FALSE)
  colnames(df) <- c("", attributes)
  return(df)
}

# Display the table using kableExtra
show_table <- function(df, weights = NULL, digits = 3) {
  if (!is.null(weights)) {
    df <- rbind(df, c("Weights", formatC(weights, digits = digits, format = "f")))
  }
  if (knitr::is_html_output()) {
    kbl(df, format = "html", escape = FALSE, align = "c") %>%
      kable_styling(full_width = FALSE, bootstrap_options = c("hover", "responsive")) %>%
      row_spec(0, background = "#f2f2f2", bold = TRUE) %>% # Bold the header row
      row_spec(nrow(df), bold = TRUE) %>% # Highlight last row
      column_spec(1, background = "#f2f2f2", bold = TRUE) %>% # Bold the first column
      column_spec(2:ncol(df))
  } else {
    kbl(df, format = "simple", escape = TRUE, align = "c")
  }
}

File: src/saaty.py
import numpy as np
from tabulate import tabulate


def round_preserve_sum(values, digits):
    # Scale factor for digits of precision
    scale = 10**digits

    # Step 1: Scale values and compute initial rounded values
    scaled = np.array(values) * scale
    floored = np.floor(scaled)

    # Total units we still need to add to match sum
    missing_units = int(round(scale - np.sum(floored)))

    # Step 2: Sort indices by largest remainder / relative error
    # We prioritize the numbers where adding 1 introduces least relative error
    indices = sorted(
        range(len(values)), key=lambda i: (floored[i] + 1 - scaled[i]) / scaled[i]
    )

    # Step 3: Distribute the missing units to minimize percent error
    result = floored.copy()
    for i in indices[:missing_units]:
        result[i] += 1

    # Step 4: Convert back to floats with desired precision
    final = result / scale
    return final.tolist()


def calculate_weights(matrix: np.ndarray, method: str = "eigenvector"):
    """
    Calculate the weights of attributes using Saaty's method.

    Args:
        matrix (np.ndarray): The Saaty matrix representing pairwise comparisons.
        method (str): The method to use for calculating weights, either 'average' or 'eigenvector'.
    Returns:
        np.ndarray: The calculated weights of the attributes.
    """
    actual_matrix = np.array(
        list(map(lambda w: 1 / (-w) if w < 0 else w, matrix.flatten()))
    ).reshape(matrix.shape)

    if method == "average":
        # Normalize the matrix
        normalized_matrix = actual_matrix / actual_matrix.sum(axis=0)
        # Calculate the average of each row
        weights = normalized_matrix.mean(axis=1)
    elif method == "eigenvector":
        # Calculate the eigenvalues and eigenvectors
        eigenvalues, eigenvectors = np.linalg.eig(actual_matrix)
        # Find the index of the maximum eigenvalue
        max_index = np.argmax(eigenvalues)
        # Get the corresponding eigenvector
        weights = np.real(eigenvectors[:, max_index])
    else:
        raise ValueError("Method must be 'average' or 'eigenvector'.")

    # Normalize the weights
    weights /= weights.sum()

    # Round the weights to preserve the sum
    weights = round_preserve_sum(weights, 3)

    return weights


def matrix_to_table(matrix: np.ndarray, attributes: list) -> list:
    """
    Convert a square matrix to a table that can be displayed in Quarto.

    Args:
        matrix (np.ndarray): The square matrix to convert.
        attributes (list): The list of attribute names corresponding to the matrix.
    Returns:
        list: A table representation of the matrix with attributes as headers.
    """
    table = []
    header = ["Attribute"] + attributes
    table.append(header)

    for i, row in enumerate(matrix):
        table_row = [attributes[i]] + list(
            map(lambda w: f"1/{-w}" if w < 0 else str(w), row.tolist())
        )
        table.append(table_row)

    return table


def show_table(table: list):
    colalign = ["left"] + ["center"] * (len(table[0]) - 1)
    return tabulate(table, headers="firstrow", tablefmt="html", colalign=colalign)


def main():

    attributes = [
        "Encasement",
        "Visibility",
        "Walking/biking accessibility to streams",
        "Walking/biking accessibility to green spaces",
        "Land use variety",
    ]

    saaty_matrix = np.array(
        [
            [1, 2, -3, -3, -2],
            [-2, 1, -4, -4, -2],
            [3, 4, 1, -2, 3],
            [3, 4, 2, 1, 1],
            [2, 2, -3, 1, 1],
        ]
    )

    # Calculate the weights
    weights = calculate_weights(saaty_matrix)

    # Convert the matrix to a table
    table = matrix_to_table(saaty_matrix, attributes)

    # Show the table
    show_table(table)


if __name__ == "__main__":
    main()
